include/construction/ctx_generic.hpp

```diff
@@ -0,0 +1,240 @@
+/*******************************************************************************
+ * include/util/ctx_generic.hpp
+ *
+ * Copyright (C) 2017 Marvin Löbel <loebel.marvin@gmail.com>
+ * Copyright (C) 2017 Florian Kurpicz <florian.kurpicz@tu-dortmund.de>
+ *
+ * All rights reserved. Published under the BSD-2 license in the LICENSE file.
+ ******************************************************************************/
+
+#pragma once
+
+#include <type_traits>
+
+#include "arrays/bit_vectors.hpp"
+#include "arrays/helper_array.hpp"
+#include "arrays/pow2_array.hpp"
+
+#include "util/permutation.hpp"
+
+namespace ctx_options {
+template <bool is_tree>
+class pre_computed_rho {
+  using rho_t = typename rho_dispatch<is_tree>::type;
+
+  rho_t const* rho_ = nullptr;
+public:
+  inline uint64_t get(size_t level, size_t i) const {
+    return (*rho_)(level, i);
+  }
+
+  inline void set(size_t /*level*/, size_t /*i*/, uint64_t /*value*/) {
+    DCHECK(false); // Should not be called
+  }
+
+  inline pre_computed_rho() = default;
+  inline pre_computed_rho(rho_t const& rho) : rho_(&rho) {}
+
+  static bool constexpr compute_rho = false;
+};
+
+template <bool is_tree>
+class live_computed_rho {
+  std::vector<uint64_t> bit_reverse_;
+public:
+  inline uint64_t get(size_t /*level*/, size_t i) const {
+    if constexpr (!is_tree) {
+      return bit_reverse_[i];
+    } else {
+      return i;
+    }
+  }
+
+  inline void set(size_t /*level*/, size_t i, uint64_t value) {
```

Comment:
```suggestion
  inline void set(size_t /*level*/, [[maybe_unused]] size_t i, [[maybe_unused]] uint64_t value) {
```
Then, we can remove  `(void) i` and `(void) value`.

Reply:
Ah right, that exists now. :D